adal/src/telemetry/java/com/microsoft/aad/adal/HttpEvent.java

```diff
@@ -88,19 +91,116 @@ void setRequestIdHeader(final String requestIdHeader) {
+     *
+     * @param xMsCliTelem the value of the x-ms-clitelem header
+     */
+    void setXMsCliTelemData(final String xMsCliTelem) {
+        // if the header isn't present, do nothing
+        if (StringExtensions.isNullOrBlank(xMsCliTelem)) {
+            return;
+        } else {
+            Logger.d(TAG, "Parsing x-ms-clitelem header: " + xMsCliTelem);
+        }
+
+        // split the header based on the delimiter
+        String[] headerSegments = xMsCliTelem.split(",");
+
+        // make sure the header isn't empty
+        if (0 == headerSegments.length) {
+            Logger.w(TAG, "SPE Ring header missing version field.", null, ADALError.X_MS_CLITELEM_VERSION_UNRECOGNIZED);
+            return;
+        }
+
+        // get the version of this header
+        final String headerVersion = headerSegments[0];
+
+        // declare values tracked by this header
+        String errorCode = null;
+        String subErrorCode = null;
+        String tokenAge = null;
+        String speRing = null;
+
+        if (headerVersion.equals("1")) {
+            // The expected delimiter count of the v1 header
+            final int delimCount = 4;
+
+            // Verify the expected format "<version>, <error_code>, <sub_error_code>, <token_age>, <ring>"
+            if (delimCount != xMsCliTelem.length() - xMsCliTelem.replace(",", "").length()) {
+                Logger.w(TAG, "", "", ADALError.X_MS_CLITELEM_MALFORMED);
+                return;
+            }
+
+            headerSegments = xMsCliTelem.split(",", 5);
+
+            final int indexErrorCode = 1;
+            final int indexSubErrorCode = 2;
+            final int indexTokenAge = 3;
+            final int indexSpeInfo = 4;
+
+            // get the error_code
+            errorCode = headerSegments[indexErrorCode];
+
+            // get the sub_error_code
+            subErrorCode = headerSegments[indexSubErrorCode];
+
+            // get the token_age
+            tokenAge = headerSegments[indexTokenAge];
+
+            // get the spe_ring
+            speRing = headerSegments[indexSpeInfo];
+        } else { // unrecognized version
+            Logger.w(TAG, "Unexpected header version: " + headerVersion, null, ADALError.X_MS_CLITELEM_VERSION_UNRECOGNIZED);
+        }
+        // Set the extracted values on the HttpEvent
+        if (!StringExtensions.isNullOrBlank(errorCode) && !errorCode.equals("0")) {
+            setSpeRingErrorCode(errorCode);
+        }
+
+        if (!StringExtensions.isNullOrBlank(subErrorCode) && !subErrorCode.equals("0")) {
+            setSpeRingSubErrorCode(subErrorCode);
+        }
+
+        if (!StringExtensions.isNullOrBlank(tokenAge)) {
+            setSpeRingTokenAge(tokenAge);
+        }
+
+        if (!StringExtensions.isNullOrBlank(speRing)) {
+            setSpeRing(speRing);
+        }
+    }
+
+    void setSpeRingErrorCode(final String errorCode) {
+        setProperty(EventStrings.SERVER_ERROR_CODE, errorCode.trim());
```

Comment:
Not important: Would also suggest renaming "setSpeRingErrorCode" to e.g setServerErrorCode as this has nothing to do with SPE ring (also true for setSpeRingSubErrorCode and setSpeRingTokenAge).

Reply:
Resolved in 7053a0912c4435694ce6d73302a2891d8153fe81